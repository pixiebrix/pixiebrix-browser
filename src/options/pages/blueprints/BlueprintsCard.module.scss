/*!
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

.root {
  // 310 is a rough estimation of page height after the headers and footers on this page.
  // Ideally we would implement a solution that doesn't use hard-coding, like flexbox
  max-height: calc(100vh - 310px);
  overflow-y: scroll;

  // TODO: this is to add padding between the div content and the scrollbar
  //  however, this messes with the content alignment with the toolbar above
  padding-right: 15px;

  &::-webkit-scrollbar {
    width: 10px;
  }

  &::-webkit-scrollbar-thumb {
    background: rgba(186, 168, 190, 0.96);
    border-radius: 10px;
  }

  &::-webkit-scrollbar-thumb:hover {
    background: rgba(169, 153, 171, 0.96);
  }
}
