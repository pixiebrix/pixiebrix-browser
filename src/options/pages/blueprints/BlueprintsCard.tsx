/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import styles from "./BlueprintsCard.module.scss";

import { Col, Row as BootstrapRow } from "react-bootstrap";
import React, { useMemo } from "react";
import {
  Column,
  useFilters,
  useGlobalFilter,
  useGroupBy,
  useSortBy,
  useTable,
} from "react-table";
import ListFilters from "./ListFilters";
import { Installable, InstallableViewItem } from "./blueprintsTypes";
import {
  selectFilters,
  selectGroupBy,
  selectSortBy,
} from "./blueprintsSelectors";
import { useSelector } from "react-redux";
import { uniq } from "lodash";
import useInstallableViewItems from "@/options/pages/blueprints/useInstallableViewItems";
import AutoSizer from "react-virtualized-auto-sizer";
import BlueprintsToolbar from "@/options/pages/blueprints/BlueprintsToolbar";
import BlueprintsView from "@/options/pages/blueprints/BlueprintsView";
import Loader from "@/components/Loader";

// These react-table columns aren't rendered as column headings,
// but used to expose grouping, sorting, filtering, and global
// searching utilities on InstallableRows
const columns: Array<Column<InstallableViewItem>> = [
  {
    Header: "Name",
    accessor: "name",
    disableGroupBy: true,
    disableFilters: true,
  },
  {
    Header: "Description",
    accessor: "description",
    disableGroupBy: true,
    disableFilters: true,
    disableSortBy: true,
  },
  {
    Header: "Package ID",
    // @ts-expect-error -- react-table allows nested accessors
    accessor: "sharing.packageId",
    disableGroupBy: true,
    disableFilters: true,
    disableSortBy: true,
  },
  {
    Header: "Source",
    // @ts-expect-error -- react-table allows nested accessors
    accessor: "sharing.source.label",
    disableGlobalFilter: true,
  },
  {
    Header: "Last updated",
    accessor: "updatedAt",
    disableGroupBy: true,
    disableFilters: true,
    disableGlobalFilter: true,
  },
  {
    Header: "Status",
    accessor: "status",
    disableGlobalFilter: true,
    filter: "exactText",
  },
];

const BlueprintsCard: React.FunctionComponent<{
  installables: Installable[];
}> = ({ installables }) => {
  const { installableViewItems, isLoading } =
    useInstallableViewItems(installables);

  const teamFilters = useMemo(
    () =>
      uniq(
        installableViewItems.map(
          (installable) => installable.sharing.source.label
        )
      ).filter((label) => label !== "Public" && label !== "Personal"),
    [installableViewItems]
  );

  const groupBy = useSelector(selectGroupBy);
  const sortBy = useSelector(selectSortBy);
  const filters = useSelector(selectFilters);

  const tableInstance = useTable<InstallableViewItem>(
    {
      columns,
      data: installableViewItems,
      initialState: {
        groupBy,
        sortBy,
        filters,
      },
      useControlledState: (state) =>
        useMemo(
          () => ({
            ...state,
            groupBy,
            sortBy,
            filters,
          }),
          // eslint-disable-next-line react-hooks/exhaustive-deps -- table props are required dependencies
          [state, groupBy, sortBy, filters]
        ),
    },
    useFilters,
    useGlobalFilter,
    useGroupBy,
    useSortBy
  );

  if (isLoading) {
    return <Loader />;
  }

  return (
    <BootstrapRow className={styles.root}>
      <ListFilters teamFilters={teamFilters} tableInstance={tableInstance} />
      <Col className={styles.mainContainer}>
        <BlueprintsToolbar tableInstance={tableInstance} />
        {/* This wrapper prevents AutoSizer overflow in a flex box container */}
        <div style={{ flex: "1 1 auto" }}>
          <AutoSizer defaultHeight={500}>
            {({ height, width }) => (
              <BlueprintsView
                tableInstance={tableInstance}
                width={width}
                height={height}
              />
            )}
          </AutoSizer>
        </div>
      </Col>
    </BootstrapRow>
  );
};

export default BlueprintsCard;
