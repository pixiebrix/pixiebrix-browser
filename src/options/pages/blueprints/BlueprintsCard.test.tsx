/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { render } from "@/options/testHelpers";
import BlueprintsCard from "@/options/pages/blueprints/BlueprintsCard";
import { Installable } from "@/options/pages/blueprints/blueprintsTypes";
import { waitForEffect } from "@/testUtils/testHelpers";
import { useGetMeQuery, useGetStarterBlueprintsQuery } from "@/services/api";
import { screen } from "@testing-library/react";
import { organizationFactory } from "@/testUtils/factories";

const EMPTY_RESPONSE = Object.freeze({
  data: Object.freeze([]),
  isLoading: false,
});

// Need to return the same object every time, because useInstallableViewItems doesn't destructure the object. Or maybe
// we just need to make sure the data [] array is the same object?
jest.mock("@/services/api", () => ({
  useGetRecipesQuery: jest.fn(() => EMPTY_RESPONSE),
  useGetMeQuery: jest.fn(() => EMPTY_RESPONSE),
  useGetCloudExtensionsQuery: jest.fn(() => EMPTY_RESPONSE),
  useGetMarketplaceListingsQuery: jest.fn(() => EMPTY_RESPONSE),
  useGetOrganizationsQuery: jest.fn(() => EMPTY_RESPONSE),
  useGetStarterBlueprintsQuery: jest.fn(() => EMPTY_RESPONSE),
}));

const installables: Installable[] = [];

describe("BlueprintsCard", () => {
  test("renders", async () => {
    const rendered = render(<BlueprintsCard installables={installables} />);
    await waitForEffect();
    expect(rendered.asFragment()).toMatchSnapshot();
  });

  test("doesn't flash the 'Get Started' tab while loading", async () => {
    (useGetStarterBlueprintsQuery as jest.Mock).mockImplementation(() => ({
      isLoading: true,
    }));

    render(<BlueprintsCard installables={installables} />);
    await waitForEffect();
    expect(
      screen.queryByText("Welcome to the PixieBrix Extension Console")
    ).toBeNull();
    expect(screen.queryByText("Get Started")).toBeNull();

    (useGetStarterBlueprintsQuery as jest.Mock).mockImplementation(() => ({
      isLoading: false,
    }));

    render(<BlueprintsCard installables={installables} />);
    await waitForEffect();
    expect(
      screen.queryByText("Welcome to the PixieBrix Extension Console")
    ).not.toBeNull();
    expect(screen.queryByText("Get Started")).not.toBeNull();
  });

  test("does not show 'Get Started' tab for enterprise users", async () => {
    (useGetMeQuery as jest.Mock).mockImplementation(() => ({
      isLoading: false,
      data: { organization: organizationFactory() },
    }));

    render(<BlueprintsCard installables={installables} />);
    await waitForEffect();
    expect(
      screen.queryByText("Welcome to the PixieBrix Extension Console")
    ).toBeNull();
    expect(screen.queryByText("Get Started")).toBeNull();
  });
});
