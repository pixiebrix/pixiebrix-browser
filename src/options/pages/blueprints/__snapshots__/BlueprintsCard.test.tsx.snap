// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlueprintsCard renders 1`] = `
<DocumentFragment>
  <div
    class="root row"
  >
    <div
      class="root col-xl-2 col-md-3 col-sm-12"
    >
      <form
        class=""
      >
        <input
          class="form-control form-control-sm"
          id="query"
          placeholder="Search all blueprints"
          value=""
        />
      </form>
      <div
        class="flex-column nav nav-pills"
      >
        <a
          class="mt-3 nav-link active"
          data-rb-event-key="Get Started"
          href="#"
          role="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-rocket fa-w-16 align-self-center"
            data-icon="rocket"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M505.12019,19.09375c-1.18945-5.53125-6.65819-11-12.207-12.1875C460.716,0,435.507,0,410.40747,0,307.17523,0,245.26909,55.20312,199.05238,128H94.83772c-16.34763.01562-35.55658,11.875-42.88664,26.48438L2.51562,253.29688A28.4,28.4,0,0,0,0,264a24.00867,24.00867,0,0,0,24.00582,24H127.81618l-22.47457,22.46875c-11.36521,11.36133-12.99607,32.25781,0,45.25L156.24582,406.625c11.15623,11.1875,32.15619,13.15625,45.27726,0l22.47457-22.46875V488a24.00867,24.00867,0,0,0,24.00581,24,28.55934,28.55934,0,0,0,10.707-2.51562l98.72834-49.39063c14.62888-7.29687,26.50776-26.5,26.50776-42.85937V312.79688c72.59753-46.3125,128.03493-108.40626,128.03493-211.09376C512.07526,76.5,512.07526,51.29688,505.12019,19.09375ZM384.04033,168A40,40,0,1,1,424.05,128,40.02322,40.02322,0,0,1,384.04033,168Z"
              fill="currentColor"
            />
          </svg>
          <span
            class="media-body ml-2"
          >
            Get Started
          </span>
        </a>
        <h5
          class="mt-3"
        >
          Category Filters
        </h5>
        <a
          class="media nav-link"
          data-rb-event-key="Active"
          href="#"
          role="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-check fa-w-16 align-self-center"
            data-icon="check"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
          <span
            class="media-body ml-2"
          >
            Active
          </span>
        </a>
        <a
          class="media nav-link"
          data-rb-event-key="All"
          href="#"
          role="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-asterisk fa-w-16 align-self-center"
            data-icon="asterisk"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M478.21 334.093L336 256l142.21-78.093c11.795-6.477 15.961-21.384 9.232-33.037l-19.48-33.741c-6.728-11.653-21.72-15.499-33.227-8.523L296 186.718l3.475-162.204C299.763 11.061 288.937 0 275.48 0h-38.96c-13.456 0-24.283 11.061-23.994 24.514L216 186.718 77.265 102.607c-11.506-6.976-26.499-3.13-33.227 8.523l-19.48 33.741c-6.728 11.653-2.562 26.56 9.233 33.037L176 256 33.79 334.093c-11.795 6.477-15.961 21.384-9.232 33.037l19.48 33.741c6.728 11.653 21.721 15.499 33.227 8.523L216 325.282l-3.475 162.204C212.237 500.939 223.064 512 236.52 512h38.961c13.456 0 24.283-11.061 23.995-24.514L296 325.282l138.735 84.111c11.506 6.976 26.499 3.13 33.227-8.523l19.48-33.741c6.728-11.653 2.563-26.559-9.232-33.036z"
              fill="currentColor"
            />
          </svg>
          <span
            class="media-body ml-2"
          >
            All Blueprints
          </span>
        </a>
        <a
          class="media nav-link"
          data-rb-event-key="Personal"
          href="#"
          role="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-user fa-w-14 align-self-center"
            data-icon="user"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
              fill="currentColor"
            />
          </svg>
          <span
            class="media-body ml-2"
          >
            Personal
          </span>
        </a>
        <a
          class="media nav-link"
          data-rb-event-key="Public"
          href="#"
          role="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-globe fa-w-16 align-self-center"
            data-icon="globe"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 496 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M336.5 160C322 70.7 287.8 8 248 8s-74 62.7-88.5 152h177zM152 256c0 22.2 1.2 43.5 3.3 64h185.3c2.1-20.5 3.3-41.8 3.3-64s-1.2-43.5-3.3-64H155.3c-2.1 20.5-3.3 41.8-3.3 64zm324.7-96c-28.6-67.9-86.5-120.4-158-141.6 24.4 33.8 41.2 84.7 50 141.6h108zM177.2 18.4C105.8 39.6 47.8 92.1 19.3 160h108c8.7-56.9 25.5-107.8 49.9-141.6zM487.4 192H372.7c2.1 21 3.3 42.5 3.3 64s-1.2 43-3.3 64h114.6c5.5-20.5 8.6-41.8 8.6-64s-3.1-43.5-8.5-64zM120 256c0-21.5 1.2-43 3.3-64H8.6C3.2 212.5 0 233.8 0 256s3.2 43.5 8.6 64h114.6c-2-21-3.2-42.5-3.2-64zm39.5 96c14.5 89.3 48.7 152 88.5 152s74-62.7 88.5-152h-177zm159.3 141.6c71.4-21.2 129.4-73.7 158-141.6h-108c-8.8 56.9-25.6 107.8-50 141.6zM19.3 352c28.6 67.9 86.5 120.4 158 141.6-24.4-33.8-41.2-84.7-50-141.6h-108z"
              fill="currentColor"
            />
          </svg>
          <span
            class="media-body ml-2"
          >
            Public Marketplace
          </span>
        </a>
      </div>
      <div
        class="flex-column nav"
      >
        <h5>
          Explore
        </h5>
        <a
          class="media nav-link"
          data-rb-event-key="https://www.pixiebrix.com/marketplace"
          href="https://www.pixiebrix.com/marketplace"
          rel="noopener noreferrer"
          target="_blank"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-external-link-alt fa-w-16 align-self-center"
            data-icon="external-link-alt"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
              fill="currentColor"
            />
          </svg>
          <span
            class="media-body ml-2"
          >
            Open Public Marketplace
          </span>
        </a>
      </div>
    </div>
    <div
      class="mainContainer col-xl-10 col-md-9 col-sm-12"
    >
      <div
        class="d-flex justify-content-between align-items-center mb-3"
      >
        <h3
          class="filterTitle"
        >
          Welcome to the PixieBrix Extension Console
        </h3>
        <span
          class="d-flex align-items-center small"
        >
          <div
            class="ml-2 css-b62m3t-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-2-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
            />
            <div
              class=" css-1s2u09g-control"
            >
              <div
                class=" css-319lph-ValueContainer"
              >
                <div
                  class=" css-14el2xx-placeholder"
                  id="react-select-2-placeholder"
                >
                  Group by
                </div>
                <div
                  class=" css-6j8wv5-Input"
                  data-value=""
                >
                  <input
                    aria-autocomplete="list"
                    aria-describedby="react-select-2-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class=""
                    id="react-select-2-input"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class=" css-1okebmr-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class=" css-tlfecz-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ml-3 css-b62m3t-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-3-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
            />
            <div
              class=" css-1s2u09g-control"
            >
              <div
                class=" css-319lph-ValueContainer"
              >
                <div
                  class=" css-14el2xx-placeholder"
                  id="react-select-3-placeholder"
                >
                  Sort by
                </div>
                <div
                  class=" css-6j8wv5-Input"
                  data-value=""
                >
                  <input
                    aria-autocomplete="list"
                    aria-describedby="react-select-3-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class=""
                    id="react-select-3-input"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class=" css-1okebmr-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class=" css-tlfecz-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <div
            class="btn-group"
            role="group"
          >
            <button
              class="ml-3 btn btn-info btn-sm"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-list fa-w-16 fa-1x "
                data-icon="list"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M80 368H16a16 16 0 0 0-16 16v64a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16v-64a16 16 0 0 0-16-16zm0-320H16A16 16 0 0 0 0 64v64a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16V64a16 16 0 0 0-16-16zm0 160H16a16 16 0 0 0-16 16v64a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16v-64a16 16 0 0 0-16-16zm416 176H176a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h320a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm0-320H176a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h320a16 16 0 0 0 16-16V80a16 16 0 0 0-16-16zm0 160H176a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h320a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <button
              class="btn btn-outline-info btn-sm"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-th-large fa-w-16 fa-1x "
                data-icon="th-large"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M296 32h192c13.255 0 24 10.745 24 24v160c0 13.255-10.745 24-24 24H296c-13.255 0-24-10.745-24-24V56c0-13.255 10.745-24 24-24zm-80 0H24C10.745 32 0 42.745 0 56v160c0 13.255 10.745 24 24 24h192c13.255 0 24-10.745 24-24V56c0-13.255-10.745-24-24-24zM0 296v160c0 13.255 10.745 24 24 24h192c13.255 0 24-10.745 24-24V296c0-13.255-10.745-24-24-24H24c-13.255 0-24 10.745-24 24zm296 184h192c13.255 0 24-10.745 24-24V296c0-13.255-10.745-24-24-24H296c-13.255 0-24 10.745-24 24v160c0 13.255 10.745 24 24 24z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </span>
      </div>
      <div
        style="flex: 1 1 auto;"
      >
        <div
          class="root"
          style="height: 600px; width: 600px;"
        >
          <div
            class="infoRow row"
          >
            <div
              class="col"
            >
              <p
                class="termDefinition"
              >
                <img
                  alt="Screenshot of the Blueprints tab in the sidebar"
                  class="inlineFigure"
                  height="70"
                  src="test-file-stub"
                />
                On this Blueprints tab, you can manage the Blueprints you have access to.
              </p>
              <p
                class="termDefinition"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-cubes fa-w-16 fa-2x inlineFigure"
                  data-icon="cubes"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M488.6 250.2L392 214V105.5c0-15-9.3-28.4-23.4-33.7l-100-37.5c-8.1-3.1-17.1-3.1-25.3 0l-100 37.5c-14.1 5.3-23.4 18.7-23.4 33.7V214l-96.6 36.2C9.3 255.5 0 268.9 0 283.9V394c0 13.6 7.7 26.1 19.9 32.2l100 50c10.1 5.1 22.1 5.1 32.2 0l103.9-52 103.9 52c10.1 5.1 22.1 5.1 32.2 0l100-50c12.2-6.1 19.9-18.6 19.9-32.2V283.9c0-15-9.3-28.4-23.4-33.7zM358 214.8l-85 31.9v-68.2l85-37v73.3zM154 104.1l102-38.2 102 38.2v.6l-102 41.4-102-41.4v-.6zm84 291.1l-85 42.5v-79.1l85-38.8v75.4zm0-112l-102 41.4-102-41.4v-.6l102-38.2 102 38.2v.6zm240 112l-85 42.5v-79.1l85-38.8v75.4zm0-112l-102 41.4-102-41.4v-.6l102-38.2 102 38.2v.6z"
                    fill="currentColor"
                  />
                </svg>
                <span>
                  A 
                  <strong>
                    Blueprint
                  </strong>
                   is like a folder or container for a group of Extensions.
                </span>
              </p>
              <p
                class="termDefinition"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-cube fa-w-16 fa-2x inlineFigure"
                  data-icon="cube"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M239.1 6.3l-208 78c-18.7 7-31.1 25-31.1 45v225.1c0 18.2 10.3 34.8 26.5 42.9l208 104c13.5 6.8 29.4 6.8 42.9 0l208-104c16.3-8.1 26.5-24.8 26.5-42.9V129.3c0-20-12.4-37.9-31.1-44.9l-208-78C262 2.2 250 2.2 239.1 6.3zM256 68.4l192 72v1.1l-192 78-192-78v-1.1l192-72zm32 356V275.5l160-65v133.9l-160 80z"
                    fill="currentColor"
                  />
                </svg>
                <span>
                  An 
                  <strong>
                    Extension
                  </strong>
                   is a workflow that allows you to automate, customize, or change things on a webpage in your browser.
                </span>
              </p>
              <p>
                You can manage the Blueprints that you've created, Blueprints from a team you're a member of, and Blueprints you activate from the Marketplace.
              </p>
            </div>
          </div>
          <div
            class="infoRow row"
          >
            <div
              class="col"
            >
              <h4>
                Want to create a new Blueprint?
              </h4>
              <ul>
                <li>
                  Start by opening a new browser tab navigating to the webpage you'd like to modify.
                </li>
                <li>
                  Go to the PixieBrix tab via the 
                  <strong>
                    Chrome DevTools
                  </strong>
                   using 
                  <kbd>
                    Ctrl + Shift + C
                  </kbd>
                   or 
                  <kbd>
                    F12
                  </kbd>
                   and start editing your page.
                </li>
                <li>
                  Save your Blueprint in the Page Editor and you'll see it show up here as a personal Blueprint.
                </li>
              </ul>
            </div>
          </div>
          <div
            class="infoRow row"
          >
            <div
              class="col"
            >
              <h4>
                Need more help?
              </h4>
              <p>
                Visit the 
                <span>
                  <a
                    href="https://docs.pixiebrix.com/quick-start-guide"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    Quick Start Guide
                  </a>
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-external-link-alt fa-w-16 fa-xs externalLinkIcon"
                    data-icon="external-link-alt"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
                 or ask questions in the 
                <span>
                  <a
                    href="https://pixiebrixcommunity.slack.com/join/shared_invite/zt-13gmwdijb-Q5nVsSx5wRLmRwL3~lsDww#/shared-invite/email"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    Slack Community
                  </a>
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-external-link-alt fa-w-16 fa-xs externalLinkIcon"
                    data-icon="external-link-alt"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
                .
              </p>
              <p>
                 Visit the 
                <span>
                  <a
                    href="https://www.pixiebrix.com/marketplace/"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    PixieBrix Marketplace
                  </a>
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-external-link-alt fa-w-16 fa-xs externalLinkIcon"
                    data-icon="external-link-alt"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
                 for ideas.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
